climakitae/threshold_tools.py

```diff
@@ -0,0 +1,613 @@
+import numpy as np
+import pandas as pd
+import xarray as xr
+from scipy import stats
+
+import lmoments3 as lm
+from lmoments3 import distr as ldistr
+from lmoments3 import stats as lstats
+
+import matplotlib.pyplot as plt
+from matplotlib import cm
+import cartopy.crs as ccrs
+import cartopy.feature as cfeature
+import geoviews as gv
+
+import holoviews as hv
+from holoviews import opts
+import hvplot.pandas
+import hvplot.xarray
+
+from .visualize import get_frequency_plot, get_geospatial_plot
+
+#####################################################################
+
+
+def get_ams(da, extremes_type="max"):
+
+    extremes_types = ["max"]
+    if extremes_type not in extremes_types:
+        raise ValueError(
+            "invalid extremes type. expected one of the following: %s" % extremes_types
+        )
+
+    if extremes_type == "max":
+        ams = da.resample(time="A").max(keep_attrs=True)
+        ams.attrs["extreme value extraction method"] = "block maxima"
+        ams.attrs["extremes type"] = "maxima"
+        ams.attrs["block size"] = "1 year"
+        ams.attrs["timeseries type"] = "annual max series"
+
+    # if extremes_type == 'min':
+    #     ams = da.resample(time='A').min(keep_attrs=True)
+    #     ams.attrs['extreme value extraction method'] = 'block maxima'
+    #     ams.attrs['extremes type'] = 'minima'
+    #     ams.attrs['block size'] = '1 year'
+    #     ams.attrs['timeseries type'] = 'annual min series'
+
+    return ams
+
+
+#####################################################################
+
+
+def get_lmoments(ams, distr="gev", multiple_points=True):
+
+    distrs = ["gev", "gumbel", "weibull", "pearson3", "genpareto"]
+    if distr not in distrs:
+        raise ValueError(
+            "invalid distr type. expected one of the following: %s" % distrs
+        )
+
+    ams_attributes = ams.attrs
+
+    if multiple_points == True:
+        ams = ams.stack(allpoints=["x", "y"]).squeeze().groupby("allpoints")
+
+    def apply_function(ams):
+        def function(ams):
+
```

Comment:
lmoments is preferred for small samples - so makes sense for say 30 years of a single run... if we are going to be applying this to multiple models, multiple ensemble members it's worth considering the option of maximum-likelihood estimators, which have the additional benefit of being able to handle (potentially) non stationary distributions... motion to add this to backlog

Reply:
Made a card for this in trello.